### PR TITLE
C17 Whales - Grace Park - Viewing Party

### PR DESCRIPTION
**Pull request by gpark439**
Could use some refactoring but passes the tests! Wave 3 took longer than all other waves combined for me. Still experiencing issues with VSCode debugger but still able to do my own workarounds right now. Thank you!!!

### CODE REVIEW

File: tests/test_wave_03.py [hunk 1]
```diff
@@ -57,9 +57,12 @@ def test_friends_unique_movies_not_duplicated():
 
     # *************************************************************************************************
     # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
+
+    assert FANTASY_2 not in friends_unique_movies
```
Comment (goeunpark) [inline tests/test_wave_03.py:61]: I would like to see more assertions here to truly test that `friend_unique_movies` do not have any duplicates, as it describes doing so on L47. It's customary for our test definition to describe in full detail what our test block intends to test. Currently, this assertion only says that `FANTASY_2` is not in friends_unique_movies, it doesn't actually test that friends_unique_movies has three distinct movies!

File: viewing_party/party.py [hunk 2]
```diff
@@ -1,23 +1,155 @@
...
+from lib2to3.pgen2.tokenize import generate_tokens
+from operator import itemgetter
+from xml.sax.handler import EntityResolver
+
+from tests.test_constants import GENRE_1, MOVIE_TITLE_1, RATING_1, USER_DATA_2
+
+# adds movie data as item in movie dictionary
 def create_movie(title, genre, rating):
-    pass
+    if title == None or genre == None or rating == None:
```
Comment (goeunpark) [inline viewing_party/party.py:12]: Great validation check from the get-go!

File: viewing_party/party.py [hunk 3]
```diff
@@ -1,23 +1,155 @@
...
+        movie = {
+            "title" : title,
+            "genre" : genre,
+            "rating" : rating
+    }
+    return movie
+
+# adds movie to watched list
+def add_to_watched(user_data, movie):
+    user_data["watched"] = [movie]
```
Comment (goeunpark) [inline viewing_party/party.py:24]: Something to consider: What will happen if `user_data["watched"]` here or `user_data["watchlist"]` on L29 contains movies already?

Instead of reassigning and overriding possibly pre-existing values here, I would look into `.append()`!

File: viewing_party/party.py [hunk 4]
```diff
@@ -1,23 +1,155 @@
...
+    user_data["watchlist"] = [movie]
+    return user_data
+
+# moves watchlist movie to watched list
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        for category, item in movie.items():
+            if category == "title" and item == title:
+                user_data["watchlist"].remove(movie)
+                user_data["watched"].append(movie)
```
Comment (goeunpark) [inline viewing_party/party.py:38]: Interesting use of deconstructing the movie dictionary into a list of tuples via `.items()` function! While this totally works, this logic can be simplified:

```
for movie in user_data["watchlist"]:
   if movie["title"] == title:
        user_data["watchlist"].remove(movie)
        user_data["watched"].append(movie)
```

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,155 @@
...
+    ratings = []
+    if len(user_data["watched"]) > 0:
+        for movie in user_data["watched"]: 
+            for category, item in movie.items():
+                if category == "rating":
+                    ratings.append(item)   
+        average_rating = ((sum(ratings))/(len(ratings)))
+    else:
+        average_rating = 0
+    return average_rating
```
Comment (goeunpark) [inline viewing_party/party.py:56]: A quick note here about variable scope in Python: 

The variable `avgerage_rating` is defined in L53 and L55 inside conditional blocks and the program accesses this variable outside of this if block in L56. This works in Python because the smallest local scope for variables are within a function. However, this will not work in many other languages (including Javascript, which we'll learn in Unit 3!) because the scope of this `average_rating` would be within the block!

Given that neat quirk, it's often good practice to define variables at its broadest "layer"! I would personally definine `avgerage_rating` at the top of the function, around L 7 to avoid this possible confusion!

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,155 @@
...
+                    unique_watched.append(user_movie)
+    return unique_watched
+
+# returns unique friends movie list from friends' movies
+def get_friends_unique_watched(user_data):
+    friends_unique_watched = []
+    user_watched = user_data["watched"]
+    friends = user_data["friends"]
+    friend_movies = []
+    for friend in friends: # loops through friend 0 and friend 1
```
Comment (goeunpark) [inline viewing_party/party.py:98]: This type of syntax (`for thing in things`) is extremely Pythonic and clear to read! Perfection!

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,155 @@
...
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+# return recommendations list from friends' movie list
+# for movies matching user's favorite genre
+def get_new_rec_by_genre(user_data):
+    recs_by_genre = []
+    favorite_genre = get_most_watched_genre(user_data)
+    recommendations = get_available_recs(user_data)
```
Comment (goeunpark) [inline viewing_party/party.py:136]: Great use of helper functions here!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,155 @@
...
+    genres = {}
+    if user_data["watched"]:
+        for movie in user_data["watched"]: 
+            for category, item in movie.items():
+                if category == "genre":
+                    if item not in genres:
+                        genres[item] = 0
+                    else:
+                        genres[item] += 1
+        popular_genre = (max(genres, key=genres.get))
```
Comment (goeunpark) [inline viewing_party/party.py:69]: Nice implementation of a counter and efficient use of the `max()` function here!

**Review by goeunpark — COMMENTED** [review]
Great job, Grace! 🎉 

Your tests pass, your commit messages are detailed and consistent, and your code swoops through nested data structures! This project is a Green. 🟢 

I've left a few comments about how you could consider DRY-ing up your code for future projects. Particularly, iterating over key-pairs using `.items()` works fine, but since we often know what specific key we're interested in, a cleaner way to do it is to index directly (see comment under L35 of `party.py` for an example). 

I also appreciated your note on this PR! Let me know if you want to go over some more VS debugging or if you have any questions about this feedback in our 1:1. Overall, this is great work -- keep it up! 💯
